LG-3711: Migrate ad hoc alerts to USWDS alert partial

aduth opened the pull request:
Related: LG-3711
Previously: #4375

**Why**: Consolidate to single, consistent alert styling.

I've had this branch sitting locally for some time, initially as an exploration of effort involved in migrating the remainder of ad-hoc alert usage.

A few notable findings:

- `alert-thumb` doesn't map directly to any [equivalent design system variant](https://design.login.gov/components/alerts/), but I assume it should be eliminated in favor of `usa-alert--success`.
- ~Following the logic, there may be an issue prior to these changes where SP custom alert texts may not be showing correctly in non-English locales. Needs more investigation.~ **Edit:** See https://github.com/18F/identity-idp/pull/4418#issuecomment-734886527.
- ~Not included here (yet): Simple Form may require more work to customize, since with the current configuration it assumes a particular alert markup which is not aligned to the USWDS alerts~
  - **Edit:** Simple Form now customized. See https://github.com/18F/identity-idp/pull/4418#issuecomment-734920675 and following comments.
  - https://github.com/18F/identity-idp/blob/b72d6150b06a1a0d26c136097933e1527b3f8925/config/initializers/simple_form.rb#L6-L7

zachmargolis commented on app/javascript/packs/email-validation.js:
```diff
@@ -5,16 +5,6 @@ function emailValidation() {
   const email = document.querySelector('input[type="email"]');
   let blurTimer;
 
-  function hideElem(elem) {
-    elem.classList.add('hide');
-    elem.hidden = true;
-  }
-
-  function showElem(elem) {
-    elem.classList.remove('hide');
-    elem.hidden = false;
-  }
```
I had to add the explicit `hidden` attribute in #3773 because of IE + JAWS accessibility reports, I think we may need to keep it

aduth replied:
> I had to add the explicit `hidden` attribute in #3773 because of IE + JAWS accessibility reports, I think we may need to keep it

I'll double-check, but if I recall correctly, I discovered that the `hide` class is implemented as something of a screen-reader-only class ([`usa-sr-only` in USWDS](https://designsystem.digital.gov/documentation/accessibility/)), which would explain why it was being read to screen readers.

<img src="https://user-images.githubusercontent.com/1779930/99676758-e56aef80-2a46-11eb-894a-a8cd6d9247b2.png" width="256" alt="hide class">

I revised this to use `display-none` instead, which hides it both visually and from screen readers.

>`display:none` [...] styles will hide content from all users. The content is removed from the visual flow of the page and is ignored by screen readers

https://webaim.org/techniques/css/invisiblecontent/#techniques

zachmargolis replied:
Yeah maybe the issue was that the `.hide` style didn't use `display: none`, I don't think I dug in that far at the time, thanks for checking! Just wanted to make sure we don't regress and get another issue filed against us

aduth replied:
> I don't think I dug in that far at the time

To be fair, it's also not the greatest choice of name, since "hide" does seem to imply `display: none`. At least, that was my initial interpretation seeing it for the first time.

aduth replied:
I was able to test this as working successfully in Safari + VoiceOver. For lack of access to JAWS, I did at least find this reference chart that appears to show that describes the expected behavior, and that JAWS abides by it:

https://stevefaulkner.github.io/HTML5accessibility/tests/hidden-2016.html

Specifically items 3 and 4:

>**4. CSS off screen** content hidden (in supporting browsers) and announced by screen reader

(Current implementation, using `hide` class)

>**3. CSS display:none** content hidden (in supporting browsers) and ignored by screen reader

(Proposed implementation, using `display-none` class)

aduth commented:
> * Following the logic, there may be an issue prior to these changes where SP custom alert texts may not be showing correctly in non-English locales. Needs more investigation.

In further testing, this didn't prove to be a problem after all. The initial worry was the difference in how we consider paths between...

https://github.com/18F/identity-idp/blob/97fe583ff97ba1a8e6c43b6bafda160b9b9cc6e4/app/decorators/service_provider_session_decorator.rb#L126-L127

https://github.com/18F/identity-idp/blob/97fe583ff97ba1a8e6c43b6bafda160b9b9cc6e4/app/views/shared/_sp_alert.html.erb#L4

But it appears that by the time the view is rendered, the `new_user_session_path` takes current locale into consideration.

aduth commented:
Updates:

- The SimpleForm customization we need seems to be only achievable by monkey-patch (6518e3c, a783a7aef). I landed at this after a thorough exploration of [the relevant source](https://github.com/heartcombo/simple_form/blob/master/lib/simple_form/error_notification.rb) revealed no additional options for markup customization, and in referencing [relevant third-party CSS framework integration documentation](https://github.com/heartcombo/simple_form/wiki/Zurb's-Foundation-integration) suggesting a similar approach. A primary alternative here is to avoid using `error_notification` altogether, if I'm correct in assuming there's no magic error notification display logic and that we're only using it in a single form ([source](https://github.com/18F/identity-idp/blob/97fe583ff97ba1a8e6c43b6bafda160b9b9cc6e4/app/views/users/passwords/edit.html.erb#L13)).
- Remove `_alert.scss` (48ba7c554) 🎉. I'd have liked to remove all of the alert icon images, but it appears we use some of them elsewhere. I think those should be updated at some point to use the design system icons, so we have consistency.

zachmargolis commented on config/initializers/simple_form.rb:
```diff
@@ -1,10 +1,10 @@
 # rubocop:disable Metrics/BlockLength
 SimpleForm.setup do |config|
+  require Rails.root.join('lib/simple_form/error_notification')
```
I *think* Rubocop when run on the server might get mad at this and ask for `Rails.root.join('lib', 'simple_form', 'error_notification')` but if it doesn't then NBD

aduth replied:
> I _think_ Rubocop when run on the server might get mad at this and ask for `Rails.root.join('lib', 'simple_form', 'error_notification')` but if it doesn't then NBD

👍  Updated in 8ba84fa.

zachmargolis reviewed (approved):
LGTM

aduth commented on lib/simple_form/error_notification.rb:
```diff
@@ -0,0 +1,11 @@
+module SimpleForm
+  class ErrorNotification
```
> Can we drop a comment describing that is a monkeypatch and not us defining our own class? Otherwise it looks like we're defining a class that's missing a bunch of other methods

Yep, good call. Added in ce8f261.

aduth commented:
Fortunately the failing tests related to outdated selectors in ec77b7f2a clued me into a regression I'd accidentally introduced, where previously we relied on the decorator's `sp_msg` method to interpolate some common variables into the message. Since I adapted `custom_alert` to return the actual alert message, this wasn't happening anymore. Rather than worrying about keeping them in sync, I removed `sp_msg` and `generate_custom_alert`, and just made `custom_alert` return the formatted text.

Updated in 313f1c4e0.

It also prompted me to figure out why my changes to `config/service_providers.yml` weren't taking effect (needed `make setup`) so that I could confirm it actually works locally: 

![Screen Shot 2020-11-27 at 1 49 10 PM](https://user-images.githubusercontent.com/1779930/100478383-07084e80-30b9-11eb-855e-d7c4e124a903.png)

aduth commented:
I was having some trouble with tests in the previous implementation having the monkey-patch at `lib/simple-form`. It appeared that the very presence and `require` of the file was preventing the base class from being defined at all, where the intention was to extend it. Loosely following the convention of [Active Support Core Extensions](https://edgeguides.rubyonrails.org/active_support_core_extensions.html) (`lib/core_extensions`), it appears to be happier at `lib/extensions/simple-form`.

I might very well be missing something obvious about load order / auto-loading / monkey-patching. Happy to adjust if I am, or if there might be a better place to put these files than creating a new directory.
